### PR TITLE
Pass --dry-run to SLiM engine to permit stronger unit tests.

### LINKED ISSUE #466: add a `dryrun` flag to the slim script for unit testing the slim engine
The generated SLiM script should check for this flag at the end of the first generation and exit if the flag is set. This flag may be set by stdpopsim unit tests for quickly checking the validity of the generated SLiM code.

### PR DESCRIPTION
**Pull request by grahamgower**
Closes #466.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Good idea - what do you see on the output as a CLI user?

**Comment by grahamgower** [comment]
Quiet.
```
$ stdpopsim -e slim HomSap 10 -o /dev/null --dry-run -q
```

Normal.
```
$ stdpopsim -e slim HomSap 10 -o /dev/null --dry-run
Simulation information:
    Engine: slim (3.3.2)
    Model id: PiecewiseConstant
    Model desciption: Piecewise constant size population model over multiple epochs.
    Population: number_samples (sampling_time_generations):
        pop0: 10 (0)
Contig Description:
    Contig length: 249250621.0
    Mean recombination rate: 1.1485597641285933e-08
    Mean mutation rate: 1.29e-08
    Genetic map: None

If you use this simulation in published work, please cite:
[simulation engine]
  Haller et al., 2019: https://doi.org/10.1111/1755-0998.12968
[genome assembly]
  The Genome Sequencing Consortium, 2001: http://dx.doi.org/10.1038/35057062
[mutation rate]
  Tian, Browning, and Browning, 2019: https://doi.org/10.1016/j.ajhg.2019.09.012
[recombination rate]
  The International HapMap Consortium, 2007: https://doi.org/10.1038/nature06258
[population size]
  Takahata, 1993: https://doi.org/10.1093/oxfordjournals.molbev.a039995
[generation time]
  Tremblay and Vezina, 2000: https://doi.org/10.1086/302770
```

Verbose.
```
$ stdpopsim -v -e slim HomSap 10 -o /dev/null --dry-run
2020-04-17 11:08:22,933 [4177195] INFO stdpopsim.cli: Running simulation model PiecewiseConstant for HomSap on Contig(length=2.5E+08, recombination_rate=1.1E-08, mutation_rate=1.3E-08, genetic_map=None) with 10 samples using slim.
Simulation information:
    Engine: slim (3.3.2)
    Model id: PiecewiseConstant
    Model desciption: Piecewise constant size population model over multiple epochs.
    Population: number_samples (sampling_time_generations):
        pop0: 10 (0)
Contig Description:
    Contig length: 249250621.0
    Mean recombination rate: 1.1485597641285933e-08
    Mean mutation rate: 1.29e-08
    Genetic map: None

// Initial random seed:
43359094502943

// RunInitializeCallbacks():
initializeTreeSeq();
initializeMutationRate(0);
initializeMutationType(1, 0.5, "f", 0);
initializeGenomicElementType(1, m1, 1);
initializeGenomicElement(g1, 0, 249250620);
initializeRecombinationRate(1.14856e-08, 249250620);

// Starting run at generation <start>:
1

If you use this simulation in published work, please cite:
[simulation engine]
  Haller et al., 2019: https://doi.org/10.1111/1755-0998.12968
[genome assembly]
  The Genome Sequencing Consortium, 2001: http://dx.doi.org/10.1038/35057062
[mutation rate]
  Tian, Browning, and Browning, 2019: https://doi.org/10.1016/j.ajhg.2019.09.012
[recombination rate]
  The International HapMap Consortium, 2007: https://doi.org/10.1038/nature06258
[population size]
  Takahata, 1993: https://doi.org/10.1093/oxfordjournals.molbev.a039995
[generation time]
  Tremblay and Vezina, 2000: https://doi.org/10.1086/302770
```

File: stdpopsim/cli.py [hunk 1]
```diff
@@ -432,6 +432,9 @@ def run_simulation(args):
             summarise_usage()
             if ts is not None:
                 write_output(ts, args)
+        elif args.dry_run and args.engine == "slim":
```
Comment (jeromekelleher) [inline stdpopsim/cli.py:435]: Potentially a little brittle here? Maybe we should just pass the ``dry_run`` argument through to all the engines and deal with it there.  Then we call ``summarise_usage`` regardless.

Comment (grahamgower) [reply stdpopsim/cli.py]: Yep, that makes loads of sense! In the msprime engine, I've passed `end_time=0` to `msprime.simulate()`. I hope this is a reasonable approach. At the very least, there should be some more thorough parameter value/type checking.

**Comment by jeromekelleher** [comment]
Very nice, that's a really good dry-run now. Just before we merge, is there any argument for returning the empty tree sequence here? Msprime would work fine - would it make sense to return the partial tree sequence from SLiM? That way we could check the populations, metadata etc are all set up correctly in the output without actually running any simulation.

Or maybe not - I don't have a strong sense of what the right thing to do is here.

**Comment by grahamgower** [comment]
It's a simple change to output the tree sequence for the initial SLiM generation, but I don't know how useful that will be. It won't contain anything about the sampling populations, because they won't have been sampled yet. But maybe there's other info that could be checked? Certainly this will confuse the user, so it shouldn't be output when `--dry-run` is used at the CLI.

```
$ stdpopsim -v -e slim HomSap -d AncientEurasia_9K19 10 20 30 40 50 60 70 -o /tmp/slim-dry-run-AncientEurasia_9K19.trees --dry-run
...
$ tskit info /tmp/slim-dry-run-AncientEurasia_9K19.trees
sequence_length:  249250621.0
trees:            1
samples:          36400
individuals:      18200
nodes:            36400
edges:            0
sites:            0
mutations:        0
migrations:       0
populations:      4
provenances:      2
```

**Comment by jeromekelleher** [comment]
Yeah, OK, it would be a gotcha. Better to not output anything.

OK, I'm sold. @petrelharp, you want to weigh in here?

**Comment by petrelharp** [comment]
Looks great - and, I agree, don't output the empty tree sequence.  There's other ways to get a simple test tree sequence quickly (large scaling factor, say).

**Comment by petrelharp** [comment]
Sounds like we should merge - I'm doing it.

**Pull request merged**